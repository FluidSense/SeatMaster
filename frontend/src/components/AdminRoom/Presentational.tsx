import AlertStripe from 'nav-frontend-alertstriper';
import HovedKnapp from 'nav-frontend-knapper';
import { Input, Textarea } from 'nav-frontend-skjema';
import { Sidetittel } from 'nav-frontend-typografi';
import React, { ChangeEvent, SyntheticEvent } from 'react';
import { ETIKETT_WARNING } from '../commonConstants';
import {
  _ALERT_CREATED_MESSAGE,
  _BUTTON_CREATE_ROOM,
  _BUTTON_DELETE_ROOM,
  _BUTTON_UPDATE_ROOM,
  _INPUT_LABEL_NAME,
  _INPUT_LABEL_NOTES,
  _TITLE_CREATE_NEW_ROOM,
  _TITLE_UPDATE_NEW_ROOM,
} from './strings';

interface IProps {
  roomName: string;
  roomNotes: string;
  setNotes: (roomNotes: ChangeEvent<HTMLInputElement>) => void;
  setName: (roomName: ChangeEvent<HTMLInputElement>) => void;
  buttonDisabled: boolean;
  onClick: () => void;
  deleteRoom: () => void;
  showAlert: boolean;
  alertMessage?: string;
  roomExists: boolean;
}

const Presentational: React.FunctionComponent<IProps> = (props) => {
  const {
    roomName,
    roomNotes,
    roomExists,
    setNotes,
    setName,
    buttonDisabled,
    onClick,
    deleteRoom,
    showAlert,
    alertMessage,
  } = props;
  const displayAlert =
    showAlert
      ? (
        <AlertStripe solid={true} type={ETIKETT_WARNING}>
          {alertMessage}
        </AlertStripe>)
      : null;
  const titleText = roomExists ? _TITLE_UPDATE_NEW_ROOM : _TITLE_CREATE_NEW_ROOM;
  const buttonText = roomExists ? _BUTTON_UPDATE_ROOM : _BUTTON_CREATE_ROOM;
  const deleteButton =
    roomExists
      ? (
        <HovedKnapp id={'delete-room-button'} type="fare" onClick={deleteRoom}>
          {_BUTTON_DELETE_ROOM}
        </HovedKnapp>)
      : null;
  // TextArea returns the wrong type, so its type has to be forced
  const assertEventType = (event: SyntheticEvent<EventTarget, Event>) => {
    const changeEvent = event as ChangeEvent<HTMLInputElement>;
    setNotes(changeEvent);
  };

  return (
    <>
      <Sidetittel>{titleText}</Sidetittel>
      {displayAlert}
      <Input
        id={'input-room-name'}
        onChange={setName}
        value={roomName}
        label={_INPUT_LABEL_NAME}
        bredde={'XL'}
      />
      <Textarea
        id={'input-room-notes'}
        onChange={assertEventType}
        value={roomNotes}
        label={_INPUT_LABEL_NOTES}
      />
      <div id="state-buttons">
        <HovedKnapp
          id={'create-room-button'}
          type="hoved"
          disabled={buttonDisabled}
          onClick={onClick}
        >
          {buttonText}
        </HovedKnapp>
        {deleteButton}
      </div>
    </>
  );
};

export default Presentational;
