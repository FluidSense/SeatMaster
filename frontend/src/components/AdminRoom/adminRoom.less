#state-buttons {
  display: flex;
  justify-content: space-around;
}

@media only screen and (max-width: 600px) {
  #state-buttons {
    display: flex;
    flex-direction: column;
  }
}
