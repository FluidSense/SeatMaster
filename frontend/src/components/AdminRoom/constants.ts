export const UPDATE_ROOM = 'UPDATE_ROOM';
export const CREATE_ROOM = 'CREATE_ROOM';
export const DELETE_ROOM = 'DELETE_ROOM';
export const ROOM_SUCCESS = 'ROOM_SUCCESS';
export const RESET_PAGE = 'RESET_PAGE';
export const ROUTE_TO = '/admin/rooms';
