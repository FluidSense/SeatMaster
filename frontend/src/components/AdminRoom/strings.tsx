export const _INPUT_LABEL_NAME = 'Name';
export const _TITLE_CREATE_NEW_ROOM = 'Create new room';
export const _TITLE_UPDATE_NEW_ROOM = 'Update room';
export const _INPUT_LABEL_NOTES = 'Notes';
export const _BUTTON_CREATE_ROOM = 'Create room';
export const _BUTTON_UPDATE_ROOM = 'Update room';
export const _BUTTON_DELETE_ROOM = 'Delete room';
export const _ALERT_CREATED_MESSAGE = 'Could not create room, something went wrong';
export const _ALERT_UPDATED_MESSAGE = 'Could not update room, something went wrong';
export const _ALERT_DELETED_MESSAGE = 'Could not delete room, something went wrong';
