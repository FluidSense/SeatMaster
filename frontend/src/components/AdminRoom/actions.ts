import { AnyAction, Dispatch } from 'redux';
import { ThunkAction } from 'redux-thunk';
import { deleteRoom, postRoom, putRoom } from '../../API/calls';
import { IPostRoom } from '../../API/interfaces';
import { CREATE_ROOM, DELETE_ROOM, RESET_PAGE, UPDATE_ROOM } from './constants';

const roomCreated = (created: boolean) => ({
  created,
  type: CREATE_ROOM,
});

const roomUpdated = (updated: boolean) => ({
  updated,
  // tslint:disable-next-line:object-literal-sort-keys --contradicting rules--
  type: UPDATE_ROOM,
});

const roomDeleted = (deleted: boolean) => ({
  deleted,
  type: DELETE_ROOM,
});

export const createRoomAction = (room: IPostRoom):
  ThunkAction<Promise<void>, {}, {}, AnyAction> => async (dispatch: Dispatch) => {
    const result = await postRoom(room);
    if (result) dispatch(roomCreated(true));
    else dispatch(roomCreated(false));
  };

export const deleteRoomAction = (id: number):
  ThunkAction<Promise<void>, {}, {}, AnyAction> => async (dispatch: Dispatch) => {
    const result = await deleteRoom(id);
    if (result) dispatch(roomDeleted(true));
    else dispatch(roomDeleted(false));
  };

export const updateRoomAction = (data: IPostRoom, id: number):
  ThunkAction<Promise<void>, {}, {}, AnyAction> => async (dispatch: Dispatch) => {
    const result = await putRoom(data, id);
    if (result) dispatch(roomUpdated(true));
    else dispatch(roomUpdated(false));
  };

export const resetPage = () => ({ payload: undefined, type: RESET_PAGE });
