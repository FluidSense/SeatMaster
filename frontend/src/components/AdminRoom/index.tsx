import React, { ChangeEvent, Component } from 'react';
import { connect } from 'react-redux';
import { Redirect } from 'react-router';
import { ThunkDispatch } from 'redux-thunk';
import { IPostRoom } from '../../API/interfaces';
import { IStore } from '../../store';
import { IRoom } from '../ViewRooms';
import {
  createRoomAction,
  deleteRoomAction,
  resetPage,
  updateRoomAction,
} from './actions';
import './adminRoom.css';
import { ROUTE_TO } from './constants';
import Presentational from './Presentational';

interface IState {
  buttonDisabled: boolean;
  redirect: boolean;
  room: IRoom;
  showAlert: boolean;
  modalOpen: boolean;
}

interface IProps {
  location: {
    room?: IRoom;
  };
}

interface IDispatchProps {
  createRoom: (data: IPostRoom) => void;
  updateRoom: (data: IPostRoom, id: number) => void;
  deleteRoom: (id: number) => void;
  reset: () => void;
}

interface IStateProps {
  submitted?: boolean;
  error?: string;
  room: any;
}

type Props = IDispatchProps & IStateProps & IProps;

// tslint:disable-next-line:class-name
class _Container extends Component<Props, IState> {
  constructor(props: Props) {
    super(props);
    this.state = {
      buttonDisabled: true,
      modalOpen: false,
      redirect: false,
      room: {
        id: -1,
        info: '',
        name: '',
        seats: {
          count: 0,
          seats: [],
        },
      },
      showAlert: false,
    };
  }

  public componentDidUpdate = (prevProps: Props, prevState: IState) => {
    const { room, showAlert } = this.state;
    const { submitted, reset } = this.props;
    // Display alertstripe
    if (submitted === false && prevProps.submitted === undefined) {
      this.setState({ showAlert: true });
      reset();
    }
    if (showAlert) setTimeout(() => this.setState({ showAlert: false }), 5000);
    // Disable or enable create room button
    if (prevState.room.name !== room.name || prevState.room.info !== room.info) {
      if (room.name.trim() !== '' && room.name.trim() !== '') {
        this.setState({ buttonDisabled: false });
      } else {
        this.setState({ buttonDisabled: true });
      }
    }
  }

  public componentDidMount = () => {
    const { room } = this.props.location;
    if (room) {
      this.setState({
        room,
      });
    }
  }

  public render() {
    const { buttonDisabled, room, showAlert, modalOpen } = this.state;
    const { submitted, reset, error } = this.props;
    const roomExists = room.id !== -1;
    const onClick = roomExists ? this.update : this.create;

    if (submitted) {
      reset();
      return <Redirect to={ROUTE_TO} />;
    }
    return (
      <Presentational
        alertMessage={error}
        buttonDisabled={buttonDisabled}
        deleteRoom={this.delete}
        onClick={onClick}
        room={room}
        roomExists={roomExists}
        setName={this.setName}
        setNotes={this.setNotes}
        showAlert={showAlert}
        toggleModal={this.toggleModal}
        modalOpen={modalOpen}
      />
    );
  }

  private setNotes = (notesEvent: ChangeEvent<HTMLInputElement>) => {
    const roomNotes = notesEvent.target.value;
    this.setState(prevState => ({
      room: {
        ...prevState.room,
        info: roomNotes,
      },
    }));
  }

  private setName = (nameEvent: ChangeEvent<HTMLInputElement>) => {
    const roomName = nameEvent.target.value;
    this.setState(prevState => ({
      room: {
        ...prevState.room,
        name: roomName,
      },
    }));
  }

  private create = () => {
    const { info: roomNotes, name: roomName } = this.state.room;
    const { createRoom } = this.props;
    const body = { info: roomNotes, name: roomName };
    createRoom(body);
  }

  private toggleModal = () => {
    this.setState({ modalOpen: !this.state.modalOpen });
  }

  private delete = () => {
    const { room } = this.state;
    const { deleteRoom } = this.props;
    if (!room) return null;
    deleteRoom(room.id);
  }

  private update = () => {
    const { room } = this.state;
    const { info: roomNotes, name: roomName } = room;
    const { updateRoom } = this.props;
    if (!room) return null;
    const body = { info: roomNotes, name: roomName };
    updateRoom(body, room.id);
  }
}

const mapStateToProps = (state: IStore) => ({
  error: state.adminRoom.error,
  room: state.adminRoom.room,
  submitted: state.adminRoom.success,
});

const mapDispatchToProps = (dispatch: ThunkDispatch<{}, {}, any>) => ({
  createRoom: (data: IPostRoom) => dispatch(createRoomAction(data)),
  deleteRoom: (id: number) => dispatch(deleteRoomAction(id)),
  reset: () => dispatch(resetPage()),
  updateRoom: (data: IPostRoom, id: number) => dispatch(updateRoomAction(data, id)),
});

const AdminRoom = connect(
  mapStateToProps,
  mapDispatchToProps,
)(_Container);

export default AdminRoom;
