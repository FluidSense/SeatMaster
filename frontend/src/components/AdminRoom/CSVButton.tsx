import KnappBase from 'nav-frontend-knapper';
import React, { Component } from 'react';
import { CSVLink } from 'react-csv';
import formatCsv, { headers, IRoomWithUsers } from '../../utils/csvFormatter';
import { _DOWNLOAD_ROOM_AS_CSV } from './strings';

interface IProps {
  fetchRoomInfo: (roomId: number) => void;
  room: IRoomWithUsers;
}

class CSVButton extends Component<IProps> {

  public render() {
    const { room } = this.props;
    if (!room || !room.seats.count) return null;
    const data = formatCsv(room);
    return (
      <CSVLink data={data} headers={headers} separator={';'} filename={`${room.name}-users.csv`}>
        <KnappBase type="flat">{_DOWNLOAD_ROOM_AS_CSV}</KnappBase>
      </CSVLink>
    );
  }
}

export default CSVButton;
