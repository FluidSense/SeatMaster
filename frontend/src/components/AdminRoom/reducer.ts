import { AnyAction } from 'redux';
import { CREATE_ROOM, DELETE_ROOM, UPDATE_ROOM } from './constants';
import { _ALERT_CREATED_MESSAGE, _ALERT_DELETED_MESSAGE, _ALERT_UPDATED_MESSAGE } from './strings';

export interface IState {
  error?: string;
  submitted: boolean;
}

const initialState = {
  submitted: false,
};

const reducer = (
  state: IState = initialState,
  action: AnyAction,
): IState => {
  const { type, payload } = action;
  switch (type) {
    case CREATE_ROOM:
      if (payload) return { ...state, submitted: payload };
      return { ...state, error: _ALERT_CREATED_MESSAGE };
    default:
      return state;
  }
};

export default reducer;
