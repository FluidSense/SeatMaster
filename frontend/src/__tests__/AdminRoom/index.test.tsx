import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import * as React from 'react';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import AdminRoom from '../../components/AdminRoom/index';

describe('Admin room functions', () => {
  const location = { room: undefined };
  const middlewares = [thunk];
  const mockStore = configureMockStore(middlewares);
  const store = mockStore({
    adminRoom: { error: undefined, submitted: undefined },
    rooms: { rooms: [] },
  });
  const mountProvider = () => (mount(
    <Provider store={store}><AdminRoom location={location} /></Provider>));

  it('renders correctly without rooms', () => {
    const wrapper = mountProvider();
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('Updates roomName state through onchange', () => {
    const wrapper = mountProvider();
    const component = wrapper.find(AdminRoom);
    const name = 'X-Wing';
    const inputName = component.find('#input-room-name').hostNodes();
    inputName.simulate('change', { target: { value: name } });
    const state = component.children().state();
    expect(state.room.name).toEqual(name);
  });

  it('Updates roomNotes state through onchange', () => {
    const wrapper = mountProvider();
    const component = wrapper.find(AdminRoom);
    const notes = 'These are some notes, bro';
    const inputNotes = component.find('#input-room-notes').hostNodes();
    inputNotes.simulate('change', { target: { value: notes } });
    const state = component.children().state();
    expect(state.room.info).toEqual(notes);
  });

  it('Renders correctly when room as prop', () => {
    const location2 = {
      room: {
        id: 64,
        info: 'wasd',
        name: 'qwerty',
        seats: { count: 0, seats: [] },
      },
    };
    const wrapper = shallow(
      <Provider store={store}><AdminRoom location={location2} /></Provider>);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});

describe('Admin Room life cycle', () => {
  const location = { room: undefined };
  const middlewares = [thunk];
  const mockStore = configureMockStore(middlewares);
  const store = mockStore({
    adminRoom: { error: undefined, submitted: undefined },
    rooms: { room: [] },
  });

  it('Update state and check if button is enabled', () => {
    const wrapper = mount(
      <Provider store={store}>
        <AdminRoom location={location} />
      </Provider>);
    const newState = {
      ...wrapper.state(),
      roomName: 'X-Wing',
      roomNotes: 'Notes',
    };
    wrapper.setState(newState);
    const component = wrapper.find(AdminRoom);
    const state = component.children().state();
    expect(state.buttonDisabled).toEqual(true);
  });
});
