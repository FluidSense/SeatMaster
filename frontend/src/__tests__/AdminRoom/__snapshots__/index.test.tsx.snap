// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Admin room functions Renders correctly when room as prop 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "storeState": Object {
        "adminRoom": Object {
          "error": undefined,
          "submitted": undefined,
        },
        "rooms": Object {
          "rooms": Array [],
        },
      },
    }
  }
>
  <Connect(_Container)
    location={
      Object {
        "room": Object {
          "id": 64,
          "info": "wasd",
          "name": "qwerty",
          "seats": Object {
            "count": 0,
            "seats": Array [],
          },
        },
      }
    }
  />
</ContextProvider>
`;

exports[`Admin room functions renders correctly without rooms 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(_Container)
    location={
      Object {
        "room": undefined,
      }
    }
  >
    <_Container
      createRoom={[Function]}
      deleteRoom={[Function]}
      fetchRoomInfo={[Function]}
      location={
        Object {
          "room": undefined,
        }
      }
      reset={[Function]}
      updateRoom={[Function]}
    >
      <Presentational
        buttonDisabled={true}
        deleteRoom={[Function]}
        fetchRoom={[Function]}
        onClick={[Function]}
        room={
          Object {
            "id": -1,
            "info": "",
            "name": "",
            "seats": Object {
              "count": 0,
              "seats": Array [],
            },
          }
        }
        roomExists={false}
        setName={[Function]}
        setNotes={[Function]}
        showAlert={false}
      >
        <div
          className="main-content"
        >
          <div
            className="title-and-button"
          >
            <Sidetittel>
              <TypografiBase
                type="sidetittel"
              >
                <h1
                  className="typo-sidetittel"
                >
                  Create new room
                </h1>
              </TypografiBase>
            </Sidetittel>
            <CSVButton
              fetchRoomInfo={[Function]}
              room={
                Object {
                  "id": -1,
                  "info": "",
                  "name": "",
                  "seats": Object {
                    "count": 0,
                    "seats": Array [],
                  },
                }
              }
            />
          </div>
          <Input
            bredde="XL"
            id="input-room-name"
            label="Name"
            onChange={[Function]}
            value=""
          >
            <div
              className="skjemaelement"
            >
              <label
                className="skjemaelement__label"
                htmlFor="input-room-name"
              >
                Name
              </label>
              <input
                className="skjemaelement__input input--xl"
                id="input-room-name"
                onChange={[Function]}
                type="text"
                value=""
              />
              <SkjemaelementFeilmelding>
                <div
                  aria-live="assertive"
                  role="alert"
                />
              </SkjemaelementFeilmelding>
            </div>
          </Input>
          <Textarea
            id="input-room-notes"
            label="Notes"
            maxLength={2000}
            onChange={[Function]}
            tellerTekst={[Function]}
            textareaClass=""
            value=""
          >
            <div
              className="skjemaelement textarea__container"
            >
              <label
                className="skjemaelement__label"
                htmlFor="input-room-notes"
              >
                Notes
              </label>
              <div
                className="textarea--medMeta__wrapper"
              >
                <EventThrottler
                  callback={[Function]}
                  delay={100}
                  event="resize"
                >
                  <textarea
                    className="skjemaelement__input textarea--medMeta"
                    id="input-room-notes"
                    onChange={[Function]}
                    style={
                      Object {
                        "height": "30px",
                      }
                    }
                    type="text"
                    value=""
                  />
                </EventThrottler>
                <p
                  className="textarea--medMeta__teller"
                >
                  <span>
                    Du har 
                    2000
                     tegn igjen
                  </span>
                </p>
              </div>
              <SkjemaelementFeilmelding>
                <div
                  aria-live="assertive"
                  role="alert"
                />
              </SkjemaelementFeilmelding>
              <div
                aria-hidden="true"
                className="textareamirror"
              />
            </div>
          </Textarea>
          <div
            id="state-buttons"
          >
            <KnappBase
              autoDisableVedSpinner={false}
              className={null}
              disabled={true}
              htmlType="submit"
              id="create-room-button"
              inaktivKlasseVedDisabled={false}
              mini={false}
              onClick={[Function]}
              spinner={false}
              type="hoved"
            >
              <button
                className="knapp knapp--hoved knapp--disabled"
                disabled={true}
                id="create-room-button"
                onClick={[Function]}
                type="submit"
              >
                Create room
              </button>
            </KnappBase>
          </div>
        </div>
      </Presentational>
    </_Container>
  </Connect(_Container)>
</Provider>
`;
